include/rmm/mr/device/prefetch_resource_adaptor.hpp

```diff
@@ -0,0 +1,143 @@
+  prefetch_resource_adaptor& operator=(prefetch_resource_adaptor const&) = delete;
+  prefetch_resource_adaptor(prefetch_resource_adaptor&&) noexcept =
+    default;  ///< @default_move_constructor
+  prefetch_resource_adaptor& operator=(prefetch_resource_adaptor&&) noexcept =
+    default;  ///< @default_move_assignment{prefetch_resource_adaptor}
+
+  /**
+   * @briefreturn{rmm::device_async_resource_ref to the upstream resource}
+   */
+  [[nodiscard]] rmm::device_async_resource_ref get_upstream_resource() const noexcept
+  {
+    return upstream_;
+  }
+
+  /**
+   * @briefreturn{Upstream* to the upstream memory resource}
+   */
+  [[nodiscard]] Upstream* get_upstream() const noexcept { return upstream_; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws rmm::bad_alloc if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    void* ptr = upstream_->allocate(bytes, stream);
+    rmm::prefetch(ptr, bytes, rmm::get_current_cuda_device(), stream);
+    return ptr;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `ptr`
+   *
+   * @param ptr Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* ptr, std::size_t bytes, cuda_stream_view stream) override
+  {
+    upstream_->deallocate(ptr, bytes, stream);
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    if (this == &other) { return true; }
+    auto cast = dynamic_cast<prefetch_resource_adaptor<Upstream> const*>(&other);
+    if (cast == nullptr) { return upstream_->is_equal(other); }
+    return get_upstream_resource() == cast->get_upstream_resource();
+  }
+
+  Upstream* upstream_;  // the upstream resource used for satisfying allocation requests
+};
+
+/**
+ * @brief Convenience factory to return a `prefetch_resource_adaptor` around the
+ * upstream resource `upstream`.
+ *
+ * @tparam Upstream Type of the upstream `device_memory_resource`.
+ * @param upstream Pointer to the upstream resource
+ * @return The new prefetch resource adaptor
+ */
+template <typename Upstream>
+prefetch_resource_adaptor<Upstream> make_prefetch_adaptor(Upstream* upstream)
```

Comment:
question: can't this just be a constructor? Why do we need a factory function?

Reply:
Good question — I followed prior art from the statistics adaptor.

Reply:
yep. Also, we should have avoided creating another adaptor with an `Upstream` template parameter, since that is something we are trying to eliminate from the current adaptors as part of our refactoring. Unfortunately, I was not around to tell you this. 

(The Upstream parameter is not necessary, it can be type erased, and will be with resource refs. See https://github.com/rapidsai/rmm/issues/1457)

Reply:
Ugh sorry, I should have caught this. I'm not yet used to the new paradigms with resource and stream refs.

Reply:
I tried to refactor the prefetching adaptor to use resource refs but got stuck at the point where I didn't know how to implement `do_is_equal`, since refs don't have an `is_equal` method (afaict). I didn't see any other adaptors that have been refactored yet.